File: Length of Last Word.py
from typing import *

class Solution:
    def lengthOfLastWord(self, s: str) -> int:

##################################################

        #文字数をカウントする変数
        count = 0
        #スペースを判定するための変数
        space = ' '
        #一文全体のスペースをカウントする変数
        spaceCount = 0
        #次のスペースまでをカウントする変数
        countToSpace = 0
        #次のスペースが終わるまで文字数から、スペースの数を引いた数値を代入する変数
        numCharMinCountSpace = 0
        #前回のnumCharMinCountSpaceを代入する変数
        previousNCMCS = 0
        #今回の値(numCharMinCountSpace)から前回の値(previousNCMCS)を引いた数を代入する変数
        subtractCNCMCSfromPNCMCS = 0
        #一番最後に出てきた単語の文字数のみを代入する変数
        numCharLastWord = 0
        #スペースが入る手前までの文字数をカウントする変数
        numCharWord = 0

        #sを一文字ずつiに代入して繰り返す
        for i in s :
            #iが文字列だった場合、もしくはスペースだった場合はtrue
            if i == '' or ' ' :
                count += 1
                previousNCMCS = numCharMinCountSpace
                #iがspaceと同じだった場合はtrue
                #それ以外はfalse、(1)へ
                if i == space:
                    countToSpace = count
                    spaceCount += 1
                    numCharMinCountSpace = countToSpace - spaceCount
                    #numCharMinCountSpaceがpreviousNCMCSと同じでなかった場合はtrue
                    if numCharMinCountSpace != previousNCMCS:
                        subtractCNCMCSfromPNCMCS = numCharMinCountSpace - previousNCMCS
                        numCharLastWord = subtractCNCMCSfromPNCMCS                        
                #(1)
                else:
                    numCharWord = count
                    numCharLastWord = numCharWord - countToSpace
        #値を引数に返す
        return numCharLastWord

##################################################

print("スタート")
print(Solution().lengthOfLastWord("ab "),"←2のはず")
print(Solution().lengthOfLastWord("a "),"←1のはず")
print(Solution().lengthOfLastWord("   fly me   to   the moon "),"←4のはず")
assert Solution().lengthOfLastWord("   fly me   to   the moon  ") == 4
assert Solution().lengthOfLastWord("Hello   World") == 5
assert Solution().lengthOfLastWord("good morning") == 7
assert Solution().lengthOfLastWord("good afternoon ") == 9
assert Solution().lengthOfLastWord("  good evening") == 7
assert Solution().lengthOfLastWord("luffy is still joyboy") == 6
assert Solution().lengthOfLastWord("Hello") == 5
print("コンプリート!本番テストへ")

"""
        #文字数をカウントする変数
        count = 0
        #スペースを判定するための変数
        space = ' '
        #次のスペースまでをカウントする変数
        countToSpace = 0
        #一番最後に出てきた単語の文字数のみを代入する変数
        numCharLastWord = 0
        #スペースが入る手前までの文字数をカウントする変数
        numCharWord = 0

        #sを一文字ずつiに代入して繰り返す
        for i in s :
            count += 1
            #iがspaceと同じだった場合はtrue
            #それ以外はfalse、(1)へ
            if i == space:
                countToSpace = count
            #(1)
            else:
                numCharWord = count
                numCharLastWord = numCharWord - countToSpace
        #値を引数に返す
        return numCharLastWord
"""

File: Length of Last Word.java
class Solution {
    public static void main(String[] args) {
      Solution ins = new Solution();
      System.out.println(ins.lengthOfLastWord("Hello   World") == 5);
      System.out.println(ins.lengthOfLastWord("   fly me   to   the moon  ") == 4);
      System.out.println(ins.lengthOfLastWord("luffy is still joyboy") == 6);
      System.out.println(ins.lengthOfLastWord("good morning") == 7);
      System.out.println(ins.lengthOfLastWord("good afternoon ") == 9);
      System.out.println(ins.lengthOfLastWord("  good evening") == 7);
      System.out.println(ins.lengthOfLastWord("Hello") == 5);
      System.out.println("すべてTrueなら一旦成功");
    }


    public int lengthOfLastWord(String s) {
      // ここにコードを書いてください

      //カウントするための変数
      int count = 0;

      //スペースを代入するための変数
      String space = " ";

      //文字列の前後のスペースを削除するための変数
      String frontAndRearSpacesExcluded = s.strip();

      //文字列全体の文字数をカウントする変数
      int oneSentence = frontAndRearSpacesExcluded.length();

      //次のスペースまでをカウントするための変数
      int countToSpace = 0;

      //ループするたびにカウントに応じたcount番目の文字を代入するための変数
      char charCount = 0;

      //一番最後に出てきた単語の文字数のみを代入するための変数
      int numCharLastWord = 0;
      
      //文字列の文字数分まで繰り返すループ文
      for(int i = 0 ; i<frontAndRearSpacesExcluded.length(); i++){
        charCount = frontAndRearSpacesExcluded.charAt(count);

        //ループするたびにカウントを＋１する
        count += 1;

        //文字(char型)をString型に変更する
        String stringCount = String.valueOf(charCount);

        //stringCountがspaceと同じだった場合はTrue
          if(stringCount.equals(space)){

            //countToSpaceにcountを代入する
            countToSpace = count;
          }
        numCharLastWord = oneSentence - countToSpace;
      }

      //値を引数に返す
      return numCharLastWord;
    }
}